subjects/ai/numpy/audit/README.md

```diff
@@ -140,180 +193,192 @@ integers[mask] = 0
-        [1, 0, 0, 0, 0, 0, 0, 0, 1],
-        [1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=int8)
-```
+[[1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]
+ [1 1 1 1 1 1 1 1 1]]
+
+[[1 1 1 1 1 1 1 1 1]
+ [1 0 0 0 0 0 0 0 1]
+ [1 0 1 1 1 1 1 0 1]
+ [1 0 1 0 0 0 1 0 1]
+ [1 0 1 0 1 0 1 0 1]
+ [1 0 1 0 0 0 1 0 1]
+ [1 0 1 1 1 1 1 0 1]
+ [1 0 0 0 0 0 0 0 1]
+ [1 1 1 1 1 1 1 1 1]]
+
+[[ 1  2  3]
+ [ 2  4  6]
+ [ 3  6  9]
+ [ 4  8 12]
+ [ 5 10 15]]
 
-##### The solution of question 2 is not accepted if the values of the array have been changed one by one manually. The usage of the for loop is not allowed neither.
-
-Here is an example of a possible solution:
-
-```python
-        x[1:8,1:8] = 0
-        x[2:7,2:7] = 1
-        x[3:6,3:6] = 0
-        x[4,4] = 1
 ```
 
-###### For question 3, is the output the following?
+##### Check the solution for cheating like:
 
-```console
-   array([[ 1,  2,  3],
-          [ 2,  4,  6],
-          [ 3,  6,  9],
-          [ 4,  8, 12],
-          [ 5, 10, 15]], dtype=int8)
-```
-
-##### The solution of question 3 is not accepted if the values of the array have been changed one by one manually. The usage of the for loop is not allowed neither.
+- The values of the array have been changed one by one manually.
+- The usage of the for loop, which is not allowed.
+- Printing the full output given in the readme.
 
-Here is an example of a possible solution:
-
-```python
-    np.reshape(arr_1, (5, 1)) * arr_2
-```
+###### Can you confirm that there was no cheating in the solution?
 
 ---
 
 ---
 
 #### Exercise 7: NaN
 
-##### The exercise is validated if all questions of the exercise are validated
-
 ###### Without having used a for loop or having filled the array manually, is the output the following?
 
 ```console
 [[ 7.  1.  7.]
-[nan  2.  2.]
-[nan  8.  8.]
-[ 9.  3.  9.]
-[ 8.  9.  8.]
-[nan  2.  2.]
-[ 8.  2.  8.]
-[nan  6.  6.]
-[ 9.  2.  9.]
-[ 8.  5.  8.]]
+ [nan  2.  2.]
+ [nan  8.  8.]
+ [ 9.  3.  9.]
+ [ 8.  9.  8.]
+ [nan  2.  2.]
+ [ 8.  2.  8.]
+ [nan  6.  6.]
+ [ 9.  2.  9.]
+ [ 8.  5.  8.]]
 ```
 
-There are two steps in this exercise:
+---
+
+---
+
+#### Exercise 8: Wine
 
-- Create the vector that contains the grade of the first exam if available or the second. This can be done using `np.where`:
+###### Was the text file successfully loaded into a NumPy array using `genfromtxt('winequality-red.csv', delimiter=';')` and optimized for memory usage, weighing `76800` bytes or less?
 
-```python
-    np.where(np.isnan(grades[:, 0]), grades[:, 1],     grades[:, 0])
-```
+Use this in the solution to confirm:
 
-- Add this vector as third column of the array. Here are two ways:
+```Python
 
-```python
-    np.insert(arr = grades, values = new_vector, axis = 1, obj = 2)
+# Check the optimized data size and absolute differences
+optimized_size = optimized_data.nbytes
```

Comment:
I would simply check if their dataset is lower than 76 800 and avoid the first check

Reply:
done